stdlib/mutex.mli

```diff
@@ -58,3 +58,14 @@ val unlock : t -> unit
 
    @before 4.12 {!Sys_error} was not raised when unlocking an unlocked mutex
    or when unlocking a mutex from a different thread. *)
+
+val with_locked : t -> (unit -> 'a) -> 'a
```

Comment:
Note: this is the same function name that @xavierleroy proposed in https://github.com/ocaml/ocaml/pull/10801#issuecomment-987061903 , so I assume Xavier is happy with the proposed API.

Reply:
I'm happy to see this function entering the standard library, as it is an excellent idiom for programming with mutexes.

Concerning the name, `with_locked` fits the `with_<desired property>` pattern, i.e. `Mutex.with_locked m action` reads as "with mutex `m` being locked, do `action`", just like `In_channel.with_open_bin f action` reads as "with file `f` being open in binary mode, do `action`".  But maybe I am being hyper-correct and `Mutex.with_lock` would feel more natural to a native English speaker.  This is for an English native speaker to say.